File: apps/web/lib/call-analytics.ts
/**
 * Pure web call-telemetry accumulator — the SSOT for the `call:analytics`
 * payload the web reports once at hangup. Parité iOS `emitCallAnalyticsSnapshot`
 * / Android `CallAnalytics`. The web was the ONE client that never emitted this
 * telemetry (prod 2026-07-12: 100% of persisted analytics rows were iOS),
 * leaving the reliability dashboard blind to web calls.
 *
 * Pure + total (no I/O, no React). The lifecycle hook threads real signals in;
 * `buildAnalyticsPayload` renders the schema-shaped payload
 * (`socketCallAnalyticsSchema`) at teardown.
 */

import type { ConnectionQualityLevel } from '@meeshy/shared/types/video-call';

export type QualitySample = {
  readonly level: ConnectionQualityLevel;
  readonly rtt: number;
  readonly packetLoss: number;
};

export type CallAnalyticsAccumulator = {
  readonly startedAtMs: number;
  readonly connectedAtMs: number | null;
  readonly reconnectionCount: number;
  readonly samples: readonly QualitySample[];
};

export type AnalyticsPayload = {
  callId: string;
  setupTimeMs: number;
  durationSeconds: number;
  reconnectionCount: number;
  networkTransitions: number;
  averageRtt: number;
  averagePacketLoss: number;
  maxPacketLoss: number;
  codec: string;
  effectsUsed: string[];
  filtersUsed: boolean;
  transcriptionUsed: boolean;
  qualityDistribution: { excellent: number; good: number; fair: number; poor: number };
  platform: string;
  deviceModel: string;
  isVideo: boolean;
  endReason: string;
};

export function createCallAnalytics(startedAtMs: number): CallAnalyticsAccumulator {
  return { startedAtMs, connectedAtMs: null, reconnectionCount: 0, samples: [] };
}

/** First connect anchors setupTimeMs; a later re-connect (post-reconnect) never re-anchors. */
export function markConnected(
  acc: CallAnalyticsAccumulator,
  nowMs: number
): CallAnalyticsAccumulator {
  if (acc.connectedAtMs !== null) return acc;
  return { ...acc, connectedAtMs: nowMs };
}

export function markReconnecting(acc: CallAnalyticsAccumulator): CallAnalyticsAccumulator {
  return { ...acc, reconnectionCount: acc.reconnectionCount + 1 };
}

export function addQualitySample(
  acc: CallAnalyticsAccumulator,
  sample: QualitySample
): CallAnalyticsAccumulator {
  return { ...acc, samples: [...acc.samples, sample] };
}

export function buildAnalyticsPayload(
  acc: CallAnalyticsAccumulator,
  opts: { callId: string; nowMs: number; isVideo: boolean; endReason: string; deviceModel?: string }
): AnalyticsPayload {
  const { samples } = acc;
  const n = samples.length;
  const setupTimeMs = acc.connectedAtMs !== null ? acc.connectedAtMs - acc.startedAtMs : -1;
  const durationSeconds = Math.max(0, Math.round((opts.nowMs - acc.startedAtMs) / 1000));

  const dist = { excellent: 0, good: 0, fair: 0, poor: 0 };
  for (const s of samples) dist[s.level] += 1;

  return {
    callId: opts.callId,
    setupTimeMs,
    durationSeconds,
    reconnectionCount: acc.reconnectionCount,
    networkTransitions: 0, // the browser has no reliable network-transition signal
    averageRtt: n > 0 ? samples.reduce((a, s) => a + s.rtt, 0) / n : 0,
    averagePacketLoss: n > 0 ? samples.reduce((a, s) => a + s.packetLoss, 0) / n : 0,
    maxPacketLoss: samples.reduce((a, s) => Math.max(a, s.packetLoss), 0),
    codec: 'unknown',
    effectsUsed: [],
    filtersUsed: false,
    transcriptionUsed: false,
    qualityDistribution: n > 0
      ? {
          excellent: dist.excellent / n,
          good: dist.good / n,
          fair: dist.fair / n,
          poor: dist.poor / n,
        }
      : { excellent: 0, good: 0, fair: 0, poor: 0 },
    platform: 'web',
    deviceModel: opts.deviceModel ?? 'web',
    isVideo: opts.isVideo,
    endReason: opts.endReason,
  };
}

File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/call/CallAnalytics.kt
package me.meeshy.sdk.model.call

/**
 * The lowercase token the `call:analytics` `endReason` field carries on the
 * wire. Spelled out explicitly (never `simpleName`) so a rename can never
 * silently change what the gateway's quality dashboards group by — same rule
 * as [ConnectionQuality.wireValue]. `Failed` deliberately drops its message:
 * the reason is a dimension, not a log line (iOS sends its enum label too).
 */
val CallEndReason.wireValue: String
    get() = when (this) {
        CallEndReason.Local -> "local"
        CallEndReason.Remote -> "remote"
        CallEndReason.Rejected -> "rejected"
        CallEndReason.Missed -> "missed"
        CallEndReason.ConnectionLost -> "connectionLost"
        is CallEndReason.Failed -> "failed"
    }

/**
 * Pure accumulator for the once-per-call `call:analytics` lifecycle telemetry
 * (iOS parity: the `CallManager` analytics accumulators folded by
 * `emitCallAnalyticsSnapshot`). Immutable — every event returns a copy, so the
 * [CallViewModel] folds it exactly like the FSM and a test can assert any
 * intermediate state.
 *
 * [fields] is the single tested SSOT of the wire payload the gateway validates
 * (`socketCallAnalyticsSchema`: every field required). `negotiationTimeMs` is
 * deliberately ABSENT — it is optional server-side and Android has no
 * answer→connected anchor yet; omitting beats sending a fake `-1` anchor.
 * Untracked-on-Android dimensions are reported as their honest zero values
 * (`networkTransitions: 0`, `codec: "unknown"`, no effects/filters/
 * transcription) rather than invented.
 */
data class CallAnalytics(
    val startedAtMs: Long,
    val connectedAtMs: Long? = null,
    val reconnectionCount: Int = 0,
    val sampleCount: Int = 0,
    val rttSumMs: Double = 0.0,
    val packetLossSum: Double = 0.0,
    val maxPacketLoss: Double = 0.0,
    val qualitySampleCounts: Map<ConnectionQuality, Int> = emptyMap(),
) {
    /** First media-up anchor; idempotent so a reconnect's re-entry never rewrites it. */
    fun connected(nowMs: Long): CallAnalytics =
        if (connectedAtMs != null) this else copy(connectedAtMs = nowMs)

    /** One ICE-restart cycle observed (entry into Reconnecting or a further attempt). */
    fun reconnecting(): CallAnalytics = copy(reconnectionCount = reconnectionCount + 1)

    /** Fold one periodic quality sample into the running aggregates. */
    fun plusSample(sample: CallQualitySample): CallAnalytics {
        val tier = ConnectionQuality.from(sample.level())
        return copy(
            sampleCount = sampleCount + 1,
            rttSumMs = rttSumMs + sample.rttMs,
            packetLossSum = packetLossSum + sample.packetLoss,
            maxPacketLoss = maxOf(maxPacketLoss, sample.packetLoss),
            qualitySampleCounts = qualitySampleCounts + (tier to ((qualitySampleCounts[tier] ?: 0) + 1)),
        )
    }

    /**
     * The complete wire payload (minus `callId`, which the transport wrapper
     * adds). `setupTimeMs` is start→first-connect, or `-1` for a call that
     * never connected (missed/rejected/failed setup) — the schema's sentinel
     * for a missing anchor.
     */
    fun fields(
        durationSeconds: Long,
        isVideo: Boolean,
        endReason: CallEndReason,
        deviceModel: String,
    ): Map<String, Any> = mapOf(
        "setupTimeMs" to (connectedAtMs?.let { it - startedAtMs } ?: -1L),
        "durationSeconds" to durationSeconds,
        "reconnectionCount" to reconnectionCount,
        "networkTransitions" to 0,
        "averageRtt" to average(rttSumMs),
        "averagePacketLoss" to average(packetLossSum),
        "maxPacketLoss" to maxPacketLoss,
        "codec" to "unknown",
        "effectsUsed" to emptyList<String>(),
        "filtersUsed" to false,
        "transcriptionUsed" to false,
        "qualityDistribution" to qualityDistribution(),
        "platform" to "android",
        "deviceModel" to deviceModel,
        "isVideo" to isVideo,
        "endReason" to endReason.wireValue,
    )

    private fun average(sum: Double): Double = if (sampleCount > 0) sum / sampleCount else 0.0

    /**
     * Fraction of samples per tier, every tier always present in [0, 1] (the
     * schema requires all four keys). Samples arrive on a fixed cadence, so
     * sample fractions approximate the time fractions iOS reports.
     */
    private fun qualityDistribution(): Map<String, Double> =
        listOf(
            ConnectionQuality.EXCELLENT,
            ConnectionQuality.GOOD,
            ConnectionQuality.FAIR,
            ConnectionQuality.POOR,
        ).associate { tier ->
            tier.wireValue to
                if (sampleCount > 0) (qualitySampleCounts[tier] ?: 0).toDouble() / sampleCount else 0.0
        }
}
